packages/vm/src/ffi.rs

```diff
@@ -1,15 +1,19 @@
 use snafu::Snafu;
 use std::fmt::Debug;
+use std::string::FromUtf8Error;
 
 /// A result type for calling into the backend via FFI. Such a call causes
 /// non-negligible computational cost and must always have gas information
 /// attached. In order to prevent new calls from forgetting such gas information
 /// to be passed, the inner success and failure types contain gas information.
-pub type FfiResult<T> = core::result::Result<FfiSuccess<T>, FfiError>;
+pub type FfiResult<T> = core::result::Result<FfiSuccess<T>, FfiFailure>;
 
 /// A return element and the gas cost of this FFI call
 pub type FfiSuccess<T> = (T, GasInfo);
 
+/// A return element and the gas cost of this FFI call
+pub type FfiFailure = (FfiError, GasInfo);
```

Comment:
Like it. GasInfo on both cases

Reply:
Would it actually make sense to return `(Result<T, FfiError>, GasInfo)`?

Then we can `process_gas_info` in all cases, then match on the result. I noticed that `process_gas_info` is called on every branch in the diff below, which opens up the chance for someone to forget one.

Reply:
It'll make the `?` operator unusable, which will make a lot of code really ugly.
We had a similar situation in SecretNetwork in a couple places, and we went with returning the gas used as an `&mut` out-parameter. This is probably nicer for making sure you don't forget about it.

Reply:
The problem here is you can do `res?` and return an error but forget to charge the gas. Notice you have to explicitly check both Ok and Err and process_gas_info on both sides. `?` is already unsafe to use.

Happy for cleaner approaches that make it easier to use `?` and also not forget to charge gas.

Reply:
`?` would only be unsafe to use if you had a conversion between `(FfiError, GasInfo)` and `VmError`, which you don't. Even in the error case the compiler forces you to unwrap the gas info, at which point you have to make the conscious choice of ignoring it.

Reply:
I considered `(Result<T, FfiError>, GasInfo)`. The biggest issue with that is Rust does not warn when you don't use this pair (and as a consequence don't handle errors).

Reply:
hmmm.... interesting, i didn't even consider that

Reply:
> I considered (Result<T, FfiError>, GasInfo). The biggest issue with that is Rust does not warn when you don't use this pair (and as a consequence don't handle errors).

Interesting point. Could you add a `#[must_use]` declaration above the type?

Reply:
I don't think so, but worth a shot.
In any case, i'm sure it'll cause a lot of noise in the code in layers that need to propagate this type without charging (not charging yet, of course)

Reply:
Oh, I was wrong. Rust finds the unused Result in the tuple.

```
warning: unused `std::result::Result` in tuple element 0 that must be used
   --> packages/vm/src/testing/mock.rs:303:9
    |
303 |         api.canonical_address(&human);
    |         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
    |
    = note: `#[warn(unused_must_use)]` on by default
    = note: this `Result` may be an `Err` variant, which should be handled

warning: 1 warning emitted
```

Reply:
> In any case, i'm sure it'll cause a lot of noise in the code in layers that need to propagate this type without charging (not charging yet, of course)

That is a good counter argument if true. Is this passed many places? afaik, it is created with gas info in the call doing the ffi and passed verbatum to the import state, where both gas and errors are handled.

Reply:
Also, I am fine merging this as is and considering other types later (maybe in another release).

Reply:
In https://github.com/CosmWasm/cosmwasm/pull/489 there is an example of `(Result<T, FfiError>, GasInfo)`. Do you like it more?